apps/drec-api/src/pods/issuer/issuer.service.ts

```diff
@@ -0,0 +1,172 @@
+import { Injectable, NotFoundException, Logger, Inject } from '@nestjs/common';
+import { Cron, CronExpression } from '@nestjs/schedule';
+import {
+  CertificateService,
+  CERTIFICATE_SERVICE_TOKEN,
+  IIssueCommandParams,
+} from '@energyweb/origin-247-certificate';
+import { CertificateSourceType, ICertificateMetadata } from '../../utils/types';
+import { DateTime } from 'luxon';
+import {
+  FilterDTO,
+  ReadsService as BaseReadsService,
+} from '@energyweb/energy-api-influxdb';
+import { DeviceService } from '../device/device.service';
+import { DeviceGroup, IDeviceGroup } from '../device-group/device-group.entity';
+import { Device, IDevice } from '../device/device.entity';
+import { BASE_READ_SERVICE } from '../reads/const';
+import { OrganizationService } from '../organization';
+import { DeviceGroupService } from '../device-group/device-group.service';
+import { ConfigService } from '@nestjs/config';
+
+@Injectable()
+export class IssuerService {
+  private readonly logger = new Logger(IssuerService.name);
+
+  constructor(
+    private groupService: DeviceGroupService,
+    private deviceService: DeviceService,
+    private organizationService: OrganizationService,
+    @Inject(CERTIFICATE_SERVICE_TOKEN)
+    private readonly certificateService: CertificateService<ICertificateMetadata>,
+    @Inject(BASE_READ_SERVICE)
+    private baseReadsService: BaseReadsService,
+    private readonly configService: ConfigService,
+  ) {}
+
+  // @Cron(CronExpression.EVERY_10_SECONDS)
+  @Cron('0 47 16 * * *') // Every day at 23:30 - Server Time
+  async handleCron(): Promise<void> {
+    const deviceGroupRule1 =
+      this.configService.get<string>('DEVICE_GROUP_RULE_1') ||
+      'registrant_organisation_code';
+    const deviceGroupRule2 =
+      this.configService.get<string>('DEVICE_GROUP_RULE_2') || 'country_code';
+    this.logger.debug('Called every day at 23:30 Server time');
+
+    const startDate = DateTime.now().minus({ days: 1 }).toUTC();
+    const endDate = DateTime.now().minus({ minute: 1 }).toUTC();
+
+    this.logger.debug(`Start date ${startDate} - End date ${endDate}`);
+
+    const groups = await this.groupService.getAll();
+    await Promise.all(
+      groups.map(async (group: DeviceGroup) => {
+        group.devices = await this.deviceService.findForGroup(group.id);
+        return await this.issueCertificateForGroup(group, startDate, endDate);
+      }),
+    );
+    const devicesWithoutGroups = await this.deviceService.findMultiple({
+      where: {
+        groupId: null,
+      },
+    });
+
+    if (devicesWithoutGroups.length) {
+      const ownerGroupedDevices = this.groupBy(
+        devicesWithoutGroups,
+        deviceGroupRule1,
+      );
+
+      await Promise.all(
```

Comment:
This actually won't work as you expect.

You got `Promise[][]` type, so nothing is awaited here. You should use `ownerGroupedDevices?.flatMap` instead of map to have `Promise[]` type

Reply:
You`re right. I don`t even need Promise.all here so I`ll just remove it.

Reply:
I think you do need await, to await `issueCertificateForGroup` method. We don't want hanging promises if possible